### PR TITLE
adds conversion methods between spec and API

### PR DESCRIPTION
**Pull request by krishchow**
This PR adds some utility methods to convert resources between the Spec and the API, these methods can, in turn, be consumed by the sidecar, or vendor drivers. There are also some minor changes to the API to make this process smoother.

### CODE REVIEW

**Comment by krishchow** [comment]
cc @wlan0

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,6 @@
 .build
 *.swp
 release-tools
+vendor
+.idea
+travis.yml
```
Comment (wlan0) [inline .gitignore:8]: travis.yml shouldn't be ignored. no?

**Review by wlan0 — CHANGES_REQUESTED** [review]
LGTM except that one comment about travis.yml

File: .gitignore (hunk 1 above)
Comment (krishchow) [reply .gitignore:8]: I'm not sure, it is ignored in the sidecar, so I added the same ignore here as well.

**Review by wlan0 — APPROVED** [review]

**Pull request merged**